Implement custom handling for HoldTap actions.

Nashenas88 opened the pull request:
Allows an API user to control the behavior of the HoldTap action
however they prefer. This allows behaviors like preventing hold
actions when rolling keys on the same side of a split keyboard.

Fixes #35

TeXitoi commented:
Great! I'll try to do a deep review in the next days.

You can avoid the CustomHandler wrapper.

Also, you can use a `fn(&Stack) -> Option<WaitingAction>` instead of the dyn types you used. it would simplify the declaration.

I have to think about making the `Stack` type public: it clearly leak implementation details, but I don't have a better idea.

TeXitoi commented on src/layout.rs:
```diff
@@ -59,7 +59,10 @@ use State::*;
+/// The current event stack.
+///
+/// Events can be retrieved by iterating over this struct and calling [Stacked::event].
+pub type Stack = ArrayDeque<[Stacked; 16], arraydeque::behavior::Wrapping>;
```
I think it would be better to avoid exposing this type directly, but only exposing some new type wrapping an iterator, allowing to change the Stack type without a breaking change.

```rust
struct StackedIter<'a>(arraydeck::Iter<'a, Stacked>);
impl<'a> Iterator for StackedIter {
    type Item = &'a Stacked;
    fn next(&mut self) -> Option<Self::Item> {
        self.0.next()
    }
    fn size_hint(&self) -> (usize, Option<usize>) {
        self.0.size_hint()
    }
}
```

Nashenas88 replied:
Use this. Thanks!

TeXitoi commented on src/layout.rs:
```diff
@@ -258,8 +273,9 @@ impl<T> WaitingState<T> {
     }
 }
 
+/// A time-tracked event.
```
```suggestion
/// An event, waiting in a stack to be processed.
```

Nashenas88 replied:
Used the updated suggestion.

TeXitoi commented on src/action.rs:
````diff
@@ -16,14 +41,65 @@ pub enum HoldTapConfig {
+    ///     tap_hold_interval: 0,
+    /// };
+    /// ```
+    Custom(CustomHandler),
````
```suggestion
    Custom(fn(StackedIter) -> Option<WaitingAction>),
```

Nashenas88 replied:
I originally attempted this, but it's not possible get rid of the `CustomerHandler` type due to trait errors on fns:
```
error: implementation of `Debug` is not general enough
  --> src/action.rs:99:12
   |
29 | #[derive(Debug, Clone, Copy, Eq, PartialEq)]
   |          ----- in this derive macro expansion
...
99 |     Custom(for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `Debug` is not general enough
   |
   = note: `Debug` would have to be implemented for the type `for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>`
   = note: ...but `Debug` is actually implemented for the type `fn(StackedIter<'0>) -> Option<WaitingAction>`, for some specific lifetime `'0`
   = note: this error originates in the derive macro `Debug` (in Nightly builds, run with -Z macro-backtrace for more info)

error: implementation of `core::cmp::Eq` is not general enough
  --> src/action.rs:99:12
   |
29 | #[derive(Debug, Clone, Copy, Eq, PartialEq)]
   |                              -- in this derive macro expansion
...
99 |     Custom(for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `core::cmp::Eq` is not general enough
   |
   = note: `core::cmp::Eq` would have to be implemented for the type `for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>`
   = note: ...but `core::cmp::Eq` is actually implemented for the type `fn(StackedIter<'0>) -> Option<WaitingAction>`, for some specific lifetime `'0`
   = note: this error originates in the derive macro `Eq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0369]: binary operation `==` cannot be applied to type `for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>`
  --> src/action.rs:99:12
   |
29 | #[derive(Debug, Clone, Copy, Eq, PartialEq)]
   |                                  --------- in this derive macro expansion
...
99 |     Custom(for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0369]: binary operation `!=` cannot be applied to type `for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>`
  --> src/action.rs:99:12
   |
29 | #[derive(Debug, Clone, Copy, Eq, PartialEq)]
   |                                  --------- in this derive macro expansion
...
99 |     Custom(for<'a> fn(StackedIter<'a>) -> Option<WaitingAction>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)
```

TeXitoi replied:
OK, that's https://users.rust-lang.org/t/why-does-not-a-function-generic-over-a-lifetime-implement-debug/17828

The "clean" solution would be to implement by hand Debug and PartialEq on Action.

Nashenas88 replied:
Just pushed. And right about to leave for vacation. Feel free to make any additional changes or wait for me to come back in a week. Thanks for the thorough reviews!

TeXitoi commented on src/action.rs:
```diff
@@ -1,6 +1,31 @@
+
+/// A newtype around a custom handler for HoldTap actions.
+#[derive(Copy, Clone)]
+pub struct CustomHandler(pub &'static (dyn Fn(&Stack) -> Option<WaitingAction> + Sync));
```
Can be avoided.

Nashenas88 replied:
See the comment above. I needed a newtype to impl Debug, PartialEq and Eq for the new variant.

TeXitoi commented on src/action.rs:
```diff
@@ -16,14 +41,65 @@ pub enum HoldTapConfig {
     PermissiveHold,
+    /// A custom configuration. Allows the behavior to be controlled by a caller
+    /// supplied handler function.
+    ///
```
Some text explaining the given input (event received after the Press event on the HoldTap, in order of appearance), and the return meaning (Some if the action is determined, None if more information is needed).

Nashenas88 replied:
Updated the docs and doc tests too. Let me know if you'd like me to reword any of it.

TeXitoi commented:
Thanks for the great contribution!

Nashenas88 commented:
I addressed all the comments so far. I'll be leaving for a 1 week vacation on Saturday, so I won't be able to respond to comments during that period. I will have time tomorrow evening to address any additional comments you might have.

TeXitoi commented on src/action.rs:
```diff
@@ -16,14 +36,76 @@ pub enum HoldTapConfig {
+    /// [Some] value will cause one of: [WaitingAction::Tap] for the configured
+    /// tap action, [WaitingAction::Hold] for the hold action, and
+    /// [WaitingAction::NoOp] to force no action to occur this cycle. A [None]
+    /// value will cause a fallback to the timeout-based approach.
```
Do you have an example when the difference between `None` and `Some(WaitingAction::NoOp)` would be interesting? Particularly, is there a case where not fallbacking to timeout is interesting?

Nashenas88 replied:
I could imagine a possible scenario where someone decides they don't want a key to activate if certain other keys are being pressed. I tend to err on the side of flexibility, but if you would prefer it without the Option that's fine too. It won't affect any of the use-cases I have for this feature.

TeXitoi reviewed (approved):
Thanks!

TeXitoi commented on src/layout.rs:
```diff
@@ -215,11 +218,18 @@ struct WaitingState<T: 'static> {
-enum WaitingAction {
+
+/// Actions that can be triggered for a key configured for HoldTap.
+#[derive(Copy, Clone, PartialEq, Eq)]
```
```suggestion
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
```

Nashenas88 commented on src/layout.rs:
```diff
@@ -215,11 +218,18 @@ struct WaitingState<T: 'static> {
+    /// Trigger the tapping event.
     Tap,
+    /// Do not trigger any event.
     NoOp,
```
Do you want to overload NoOp or keep the Option behavior? It seems to like above you preferred a "fallback" meaning to the public API and a "do nothing this tick" to the private usage. I'm ok with this. It could have a doc-comment for the public usage and a non-doc comment for the private behavior as well.

TeXitoi replied:
I don't really know what is better. My main concern is easy to use, for the library user, API. It is better to have consistent code for maintenance, having a different behavior internally and externally is dangerous.

I badly chosen `NoOp` for this case (it was a private API, so I didn't think too much on the naming), as, internally, it means "don't choose now".

I think there is several posibilities. Some of them:
* change the name `NoOp` to something like `ChooseLater` (I can't find a proper name), and the function don't return an option.
* remove `NoOp`, and None mean "chose later"
* keep `NoOp`, but change the handling to not generate any event in this case, and removing the Waiting thing. This solution seems to be want you wanted to do. The internal code must be changed to handle this case.
* maybe something else.

TeXitoi replied:
In https://github.com/TeXitoi/keyberon/pull/82#discussion_r851668067 I think you "want" the function to say "don't generate any event, just forget about it" by returning NoOp. So I was talking of this feature.

Nashenas88 replied:
Ok, I just pushed a change that includes the "don't generate an event" behavior along with an additional test in layout to ensure it works as expected.

Nashenas88 commented:
Just wanted to follow up since it's been a bit. Is there anything else I needed to address?

TeXitoi commented:
Sorry, i didn't had the time to review since last time

TeXitoi commented on src/layout.rs:
```diff
@@ -771,4 +812,111 @@ mod test {
+    #[test]
+    fn custom_handler() {
+        fn always_tap(_: StackedIter) -> Option<WaitingAction> {
+            std::println!("hit!");
```
```suggestion
```

TeXitoi commented on src/layout.rs:
```diff
@@ -771,4 +812,111 @@ mod test {
+        fn always_none(_: StackedIter) -> Option<WaitingAction> {
+            None
+        }
+        std::println!("Assigning layers");
```
```suggestion
```

TeXitoi commented:
Thanks, and sorry for the duration of the review

Nashenas88 commented:
No worries! I only re-requested the review when I realized it might have been missing from a queue of yours. No rush intended, and thanks for the thorough review!
